## Day 1

In [234]:
import pandas as pd
from scipy.io import arff

# Load the arff file
data = arff.loadarff('../data/raw/cm1.arff')

# Convert to a pandas DataFrame
df = pd.DataFrame(data[0])

print(df.head())

    loc  v(g)  ev(g)  iv(g)     n       v     l      d      i        e  ...  \
0   1.1   1.4    1.4    1.4   1.3    1.30  1.30   1.30   1.30     1.30  ...   
1   1.0   1.0    1.0    1.0   1.0    1.00  1.00   1.00   1.00     1.00  ...   
2  24.0   5.0    1.0    3.0  63.0  309.13  0.11   9.50  32.54  2936.77  ...   
3  20.0   4.0    4.0    2.0  47.0  215.49  0.06  16.00  13.47  3447.89  ...   
4  24.0   6.0    6.0    2.0  72.0  346.13  0.06  17.33  19.97  5999.58  ...   

   lOCode  lOComment  lOBlank  locCodeAndComment  uniq_Op  uniq_Opnd  \
0     2.0        2.0      2.0                2.0      1.2        1.2   
1     1.0        1.0      1.0                1.0      1.0        1.0   
2     1.0        0.0      6.0                0.0     15.0       15.0   
3     0.0        0.0      3.0                0.0     16.0        8.0   
4     0.0        0.0      3.0                0.0     16.0       12.0   

   total_Op  total_Opnd  branchCount   defects  
0       1.2         1.2          1.4  b'fal

In [235]:
print("Shape:")
print(df.shape)

print("\nColumn names:")
print(df.columns.tolist())

print("\nData types:")
print(df.dtypes)

print("\nMissing values:")
print(df.isnull().sum())

Shape:
(498, 22)

Column names:
['loc', 'v(g)', 'ev(g)', 'iv(g)', 'n', 'v', 'l', 'd', 'i', 'e', 'b', 't', 'lOCode', 'lOComment', 'lOBlank', 'locCodeAndComment', 'uniq_Op', 'uniq_Opnd', 'total_Op', 'total_Opnd', 'branchCount', 'defects']

Data types:
loc                  float64
v(g)                 float64
ev(g)                float64
iv(g)                float64
n                    float64
v                    float64
l                    float64
d                    float64
i                    float64
e                    float64
b                    float64
t                    float64
lOCode               float64
lOComment            float64
lOBlank              float64
locCodeAndComment    float64
uniq_Op              float64
uniq_Opnd            float64
total_Op             float64
total_Opnd           float64
branchCount          float64
defects               object
dtype: object

Missing values:
loc                  0
v(g)                 0
ev(g)                0
iv(g)       

In [236]:
df["defects"].value_counts()

defects
b'false'    449
b'true'      49
Name: count, dtype: int64

In [237]:
df["defects"].value_counts(normalize=True) * 100

defects
b'false'    90.160643
b'true'      9.839357
Name: proportion, dtype: float64

In [238]:
df.describe()

,loc,v(g),ev(g),iv(g),n,v,l,d,i,e,...,t,lOCode,lOComment,lOBlank,locCodeAndComment,uniq_Op,uniq_Opnd,total_Op,total_Opnd,branchCount
count,498.000000,498.000000,498.000000,498.000000,498.000000,498.000000,498.000000,498.000000,498.000000,4.980000e+02,...,498.000000,498.000000,498.000000,498.000000,498.000000,498.000000,498.000000,498.000000,498.000000,498.000000
mean,29.644779,5.382329,2.490763,3.528916,143.956426,900.175823,0.146325,15.829378,38.455361,3.488493e+04,...,1938.056124,3.787149,12.283133,11.534137,0.006024,15.199197,25.452209,88.389960,55.570683,9.348193
std,42.753572,8.347359,3.658847,5.464398,221.049888,1690.814334,0.159337,15.330960,36.996297,1.341647e+05,...,7453.591519,8.508658,25.828605,19.981476,0.100120,9.617815,33.925816,134.917513,86.969527,15.072219
min,1.000000,1.000000,1.000000,1.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000e+00,...,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,0.000000,1.000000,0.000000,1.000000
25%,8.000000,1.000000,1.000000,1.000000,25.000000,102.190000,0.050000,5.630000,16.210000,6.061700e+02,...,33.672500,0.000000,0.000000,1.000000,0.000000,9.000000,7.000000,15.000000,10.000000,1.000000
50%,17.000000,3.000000,1.000000,2.000000,67.500000,329.820000,0.090000,11.640000,27.400000,3.677620e+03,...,204.310000,1.000000,4.000000,5.000000,0.000000,14.000000,15.000000,42.000000,26.000000,5.000000
75%,31.000000,6.000000,1.000000,4.000000,151.750000,861.460000,0.177500,21.142500,46.900000,1.663334e+04,...,924.075000,4.000000,14.000000,13.000000,0.000000,20.000000,30.000000,94.750000,59.750000,11.000000
max,423.000000,96.000000,30.000000,63.000000,2075.000000,17124.280000,1.300000,125.770000,293.680000,2.153691e+06,...,119649.480000,80.000000,339.000000,164.000000,2.000000,72.000000,314.000000,1261.000000,814.000000,162.000000


## Initial EDA Findings
* The class distribution of the data set is imbalanced.
* The CM1 dataset consists of 498 rows and 22 columns. Rows contain information on software modules and columns contain attributes of teh respective software modules.
* There are 22 attribute columns out of which 21 columns are used as explanatory variates (input of the model) and 1 is response variate (output of the model).
* There are no missing values found in the dataset.
* There are total of 449 non-defective modules and 49 defective modules out of 498 software modules.
* Several explanatory variables are rightly skewed because their means are larger than their medians.

## Day 2

In [239]:
X = df.drop(columns=["defects"])
Y = df["defects"]

print(X.shape)
print(Y.shape)

(498, 21)
(498,)


In [240]:
from sklearn.model_selection import train_test_split

In [241]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42, stratify=Y)

In [242]:
print(X_train.shape)
print(X_test.shape)
print(Y_train.shape)
print(Y_test.shape)

(398, 21)
(100, 21)
(398,)
(100,)


In [243]:
Y.value_counts(normalize=True)
Y_train.value_counts(normalize=True)
Y_test.value_counts(normalize=True)

defects
b'false'    0.9
b'true'     0.1
Name: proportion, dtype: float64

## Day 3

In [244]:
from sklearn.dummy import DummyClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, precision_score, recall_score, f1_score, roc_auc_score
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import make_pipeline

In [245]:
Y_train = Y_train.astype(str)
Y_test = Y_test.astype(str)

dummy = DummyClassifier(strategy="most_frequent")

In [246]:
dummy.fit(X_train, Y_train)
dummy_prediction = dummy.predict(X_test)

print(dummy_prediction[:20])

['false' 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false'
 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false'
 'false' 'false']


In [247]:
dummyAccuracy = accuracy_score(Y_test, dummy_prediction)
dummyConfusionMatrix = confusion_matrix(Y_test, dummy_prediction)

print("Accuracy:", dummyAccuracy)
print("Confusion Matrix:")
print(dummyConfusionMatrix)

Accuracy: 0.9
Confusion Matrix:
[[90  0]
 [10  0]]


In [248]:
precision = precision_score(Y_test, dummy_prediction, pos_label="true", zero_division=0)
recall = recall_score(Y_test, dummy_prediction, pos_label="true")
f1 = f1_score(Y_test, dummy_prediction, pos_label="true", zero_division=0)

print("Precision:", precision)
print("Recall:", recall)
print("F1-score:", f1)

Precision: 0.0
Recall: 0.0
F1-score: 0.0


In [249]:
pipe = make_pipeline(StandardScaler(), LogisticRegression())

In [250]:
pipe.fit(X_train, Y_train)
pipe_prediction = pipe.predict(X_test)
pipeProbability = pipe.predict_proba(X_test)
pipe_defectiveProbability = pipeProbability[:, 1]

print(pipe_prediction[:20])

print("\nClasses:")
print(pipe.classes_)

print("\nProbability shape:")
print(pipeProbability.shape)

print("\nFirst 5 probability rows:")
print(pipeProbability[:5])

['false' 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false'
 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false'
 'false' 'false']

Classes:
['false' 'true']

Probability shape:
(100, 2)

First 5 probability rows:
[[0.96404898 0.03595102]
 [0.99060182 0.00939818]
 [0.87225939 0.12774061]
 [0.97451593 0.02548407]
 [0.96579919 0.03420081]]


In [251]:
pipeAccuracyScore = accuracy_score(Y_test, pipe_prediction)
pipeConfusionMatrix = confusion_matrix(Y_test, pipe_prediction)
pipePrecision = precision_score(Y_test, pipe_prediction, pos_label="true", zero_division=0)
pipeRecall = recall_score(Y_test, pipe_prediction, pos_label="true")
pipeF1 = f1_score(Y_test, pipe_prediction, pos_label="true", zero_division=0)
pipeROC_AUC = roc_auc_score(Y_test, pipe_defectiveProbability)

print("Accuracy: ", pipeAccuracyScore)
print("\nConfusion Matrix: ")
print(pipeConfusionMatrix)

print("\nPrecision:", pipePrecision)
print("\nRecall:", pipeRecall)
print("\nF1-score:", pipeF1)
print("\nROC-AUC Score:", pipeROC_AUC)

Accuracy:  0.9

Confusion Matrix: 
[[89  1]
 [ 9  1]]

Precision: 0.5

Recall: 0.1

F1-score: 0.16666666666666666

ROC-AUC Score: 0.6366666666666667


## Day 3 Findings

The DummyClassifier baseline achieved about 90.0% accuracy but detected 0 defective modules. Logistic Regression achieved about 90.0% accuracy and detected 1 defective modules, but it still missed 9 of the 10 defective modules in the test set. This shows that Logistic Regression improves defect detection compared with the baseline, but it still has difficulty identifying defective modules. Therefore, accuracy alone does not give a complete picture of the model's performance, especially because the dataset is imbalanced.

## Day 4

In [252]:
from sklearn.ensemble import RandomForestClassifier

In [253]:
randomForest = RandomForestClassifier(random_state=42)

In [254]:
randomForest.fit(X_train, Y_train)

randomForest_prediction = randomForest.predict(X_test)
randomForest_probability = randomForest.predict_proba(X_test)
randomForest_defectiveProbability =randomForest_probability[:, 1]

print(randomForest_prediction[:20])

print("\nClasses:")
print(randomForest.classes_)

print("\nProbability shape:")
print(randomForest_probability.shape)

print("\nFirst 5 probability rows:")
print(randomForest_probability[:5])


['false' 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false'
 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false' 'false'
 'false' 'false']

Classes:
['false' 'true']

Probability shape:
(100, 2)

First 5 probability rows:
[[0.87 0.13]
 [1.   0.  ]
 [0.83 0.17]
 [0.99 0.01]
 [1.   0.  ]]


In [255]:
randomForest_confusionMatrix = confusion_matrix(Y_test, randomForest_prediction)
randomForest_accuracy = accuracy_score(Y_test, randomForest_prediction)
randomForest_precision = precision_score(Y_test, randomForest_prediction, pos_label="true", zero_division=0)
randomForest_recall = recall_score(Y_test, randomForest_prediction, pos_label="true", zero_division=0)
randomForest_F1 = f1_score(Y_test, randomForest_prediction, pos_label="true", zero_division=0)
randomForest_ROC_AUC = roc_auc_score(Y_test, randomForest_defectiveProbability)

print("Confusion Matrix: ")
print(randomForest_confusionMatrix)
print("\nAccuracy: ", randomForest_accuracy)
print("\nPrecision:", randomForest_precision)
print("\nRecall:", randomForest_recall)
print("\nF1-score:", randomForest_F1)
print("\nROC-AUC Score:", randomForest_ROC_AUC)

Confusion Matrix: 
[[88  2]
 [ 9  1]]

Accuracy:  0.89

Precision: 0.3333333333333333

Recall: 0.1

F1-score: 0.15384615384615385

ROC-AUC Score: 0.6244444444444445


## Day 4 Findings

Logistic Regression achieved 90.0% accuracy, while Random Forest achieved 89.0% accuracy. Both models detected only 1 of the 10 defective modules, giving both models a recall of 10%. Logistic Regression had slightly higher precision, F1-score and ROC-AUC than Random Forest. Therefore, Random Forest did not outperform Logistic Regression for defect detection on CM1 under this train/test split.

Although the accuracy values were high, both models missed 9 of the 10 defective modules. This shows that accuracy alone does not provide a complete picture of defect-detection, especially with an imbalanced dataset. The small number of defective modules in the test set also means that each correctly detected defect has a large effect on recall.

## Day 5

In [256]:
from sklearn.model_selection import StratifiedKFold
from sklearn.model_selection import cross_validate
from sklearn.metrics import make_scorer, precision_score, recall_score, f1_score

In [257]:
skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

In [258]:
metric_names = {
    "accuracy": "accuracy",
    "precision": make_scorer(precision_score, pos_label="true"),
    "recall": make_scorer(recall_score, pos_label="true"),
    "f1": make_scorer(f1_score, pos_label="true"),
    "roc_auc": "roc_auc"
}

In [259]:
baseline_cv = cross_validate(estimator=dummy, X=X_train, y=Y_train, cv=skf, scoring=metric_names)
print(baseline_cv, "\n")

{'fit_time': array([0.00064301, 0.00063515, 0.00056505, 0.00063682, 0.00047088]), 'score_time': array([0.00922394, 0.00743699, 0.00582004, 0.00628901, 0.00501108]), 'test_accuracy': array([0.9       , 0.9       , 0.9       , 0.91139241, 0.89873418]), 'test_precision': array([0., 0., 0., 0., 0.]), 'test_recall': array([0., 0., 0., 0., 0.]), 'test_f1': array([0., 0., 0., 0., 0.]), 'test_roc_auc': array([0.5, 0.5, 0.5, 0.5, 0.5])} 



/Users/saiyamrajput/Desktop/Work/Software Defect Prediction/.venv/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1879: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
/Users/saiyamrajput/Desktop/Work/Software Defect Prediction/.venv/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1879: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
/Users/saiyamrajput/Desktop/Work/Software Defect Prediction/.venv/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1879: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter

In [260]:
print("Accuracy mean:", baseline_cv["test_accuracy"].mean())
print("Accuracy std:", baseline_cv["test_accuracy"].std())

print("\nPrecision mean:", baseline_cv["test_precision"].mean())
print("Precision std:", baseline_cv["test_precision"].std())

print("\nRecall mean:", baseline_cv["test_recall"].mean())
print("Recall std:", baseline_cv["test_recall"].std())

print("\nF1 mean:", baseline_cv["test_f1"].mean())
print("F1 std:", baseline_cv["test_f1"].std())

print("\nROC-AUC mean:", baseline_cv["test_roc_auc"].mean())
print("ROC-AUC std:", baseline_cv["test_roc_auc"].std())

Accuracy mean: 0.9020253164556962
Accuracy std: 0.0047091329715792905

Precision mean: 0.0
Precision std: 0.0

Recall mean: 0.0
Recall std: 0.0

F1 mean: 0.0
F1 std: 0.0

ROC-AUC mean: 0.5
ROC-AUC std: 0.0


In [261]:
logistic_cv = cross_validate(estimator=pipe, X=X_train, y=Y_train, cv=skf, scoring=metric_names)
print(logistic_cv, "\n")

{'fit_time': array([0.00612617, 0.00335383, 0.00350904, 0.0041573 , 0.00374413]), 'score_time': array([0.0089798 , 0.00635004, 0.00610614, 0.00598884, 0.00637007]), 'test_accuracy': array([0.8875    , 0.8875    , 0.8875    , 0.91139241, 0.91139241]), 'test_precision': array([0.4       , 0.        , 0.        , 0.5       , 0.66666667]), 'test_recall': array([0.25      , 0.        , 0.        , 0.14285714, 0.25      ]), 'test_f1': array([0.30769231, 0.        , 0.        , 0.22222222, 0.36363636]), 'test_roc_auc': array([0.77604167, 0.76388889, 0.76388889, 0.80753968, 0.90140845])} 



In [262]:
print("Accuracy mean:", logistic_cv["test_accuracy"].mean())
print("Accuracy std:", logistic_cv["test_accuracy"].std())

print("\nPrecision mean:", logistic_cv["test_precision"].mean())
print("Precision std:", logistic_cv["test_precision"].std())

print("\nRecall mean:", logistic_cv["test_recall"].mean())
print("Recall std:", logistic_cv["test_recall"].std())

print("\nF1 mean:", logistic_cv["test_f1"].mean())
print("F1 std:", logistic_cv["test_f1"].std())

print("\nROC-AUC mean:", logistic_cv["test_roc_auc"].mean())
print("ROC-AUC std:", logistic_cv["test_roc_auc"].std())

Accuracy mean: 0.8970569620253164
Accuracy std: 0.011704840226590511

Precision mean: 0.31333333333333335
Precision std: 0.2696499788820891

Recall mean: 0.12857142857142856
Recall std: 0.11203133672398659

F1 mean: 0.17871017871017872
F1 std: 0.15271020748329703

ROC-AUC mean: 0.8025535155376705
ROC-AUC std: 0.05194216958598094


In [263]:
randomForest_cv = cross_validate(estimator=randomForest, X=X_train, y=Y_train, cv=skf, scoring=metric_names)
print(randomForest_cv, "\n")

{'fit_time': array([0.0699451 , 0.06335902, 0.05978608, 0.05959797, 0.0600791 ]), 'score_time': array([0.0128901 , 0.01142597, 0.0110352 , 0.01094794, 0.01180601]), 'test_accuracy': array([0.875     , 0.875     , 0.875     , 0.88607595, 0.89873418]), 'test_precision': array([0., 0., 0., 0., 0.]), 'test_recall': array([0., 0., 0., 0., 0.]), 'test_f1': array([0., 0., 0., 0., 0.]), 'test_roc_auc': array([0.828125  , 0.65885417, 0.70052083, 0.82837302, 0.93397887])} 



/Users/saiyamrajput/Desktop/Work/Software Defect Prediction/.venv/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1879: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])


In [264]:
print("Accuracy mean:", randomForest_cv["test_accuracy"].mean())
print("Accuracy std:", randomForest_cv["test_accuracy"].std())

print("\nPrecision mean:", randomForest_cv["test_precision"].mean())
print("Precision std:", randomForest_cv["test_precision"].std())

print("\nRecall mean:", randomForest_cv["test_recall"].mean())
print("Recall std:", randomForest_cv["test_recall"].std())

print("\nF1 mean:", randomForest_cv["test_f1"].mean())
print("F1 std:", randomForest_cv["test_f1"].std())

print("\nROC-AUC mean:", randomForest_cv["test_roc_auc"].mean())
print("ROC-AUC std:", randomForest_cv["test_roc_auc"].std())

Accuracy mean: 0.8819620253164556
Accuracy std: 0.009419541814010897

Precision mean: 0.0
Precision std: 0.0

Recall mean: 0.0
Recall std: 0.0

F1 mean: 0.0
F1 std: 0.0

ROC-AUC mean: 0.7899703778224905
ROC-AUC std: 0.09885511255587849


In [265]:
models = ["DummyClassifier", "Logistic Regression", "Random Forest Classifier"]
meanTableAccuracy = [baseline_cv["test_accuracy"].mean(), logistic_cv["test_accuracy"].mean(), randomForest_cv["test_accuracy"].mean()]
meanTablePrecision = [baseline_cv["test_precision"].mean(), logistic_cv["test_precision"].mean(), randomForest_cv["test_precision"].mean()]
meanTableRecall = [baseline_cv["test_recall"].mean(), logistic_cv["test_recall"].mean(), randomForest_cv["test_recall"].mean()]
meanTableF1 = [baseline_cv["test_f1"].mean(), logistic_cv["test_f1"].mean(), randomForest_cv["test_f1"].mean()]
meanTableRocAuc = [baseline_cv["test_roc_auc"].mean(), logistic_cv["test_roc_auc"].mean(), randomForest_cv["test_roc_auc"].mean()]

meanComparisonTable = pd.DataFrame({
    "Model": models,
    "Accuracy Mean": meanTableAccuracy,
    "Precision Mean": meanTablePrecision,
    "Recall Mean": meanTableRecall,
    "F1 Mean": meanTableF1,
    "ROC-AUC Mean": meanTableRocAuc
})

print(meanComparisonTable)

                      Model  Accuracy Mean  Precision Mean  Recall Mean  \
0           DummyClassifier       0.902025        0.000000     0.000000   
1       Logistic Regression       0.897057        0.313333     0.128571   
2  Random Forest Classifier       0.881962        0.000000     0.000000   

   F1 Mean  ROC-AUC Mean  
0  0.00000      0.500000  
1  0.17871      0.802554  
2  0.00000      0.789970  


In [266]:
stdTableAccuracy = [baseline_cv["test_accuracy"].std(), logistic_cv["test_accuracy"].std(), randomForest_cv["test_accuracy"].std()]
stdTablePrecision = [baseline_cv["test_precision"].std(), logistic_cv["test_precision"].std(), randomForest_cv["test_precision"].std()]
stdTableRecall = [baseline_cv["test_recall"].std(), logistic_cv["test_recall"].std(), randomForest_cv["test_recall"].std()]
stdTableF1 = [baseline_cv["test_f1"].std(), logistic_cv["test_f1"].std(), randomForest_cv["test_f1"].std()]
stdTableRocAuc = [baseline_cv["test_roc_auc"].std(), logistic_cv["test_roc_auc"].std(), randomForest_cv["test_roc_auc"].std()]

stdComparisonTable = pd.DataFrame({
    "Model STD": models,
    "Accuracy STD": stdTableAccuracy,
    "Precision STD": stdTablePrecision,
    "Recall STD": stdTableRecall,
    "F1 STD": stdTableF1,
    "ROC-AUC STD": stdTableRocAuc
})

print(stdComparisonTable)

                  Model STD  Accuracy STD  Precision STD  Recall STD   F1 STD  \
0           DummyClassifier      0.004709        0.00000    0.000000  0.00000   
1       Logistic Regression      0.011705        0.26965    0.112031  0.15271   
2  Random Forest Classifier      0.009420        0.00000    0.000000  0.00000   

   ROC-AUC STD  
0     0.000000  
1     0.051942  
2     0.098855  


In [267]:
finalComparisonTable = pd.DataFrame({
    "Model": models,
    "Accuracy": [f"{meanTableAccuracy[i]:.3f} ± {stdTableAccuracy[i]:.3f}" for i in range(3)],
    "Precision": [f"{meanTablePrecision[i]:.3f} ± {stdTablePrecision[i]:.3f}" for i in range(3)],
    "Recall": [f"{meanTableRecall[i]:.3f} ± {stdTableRecall[i]:.3f}" for i in range(3) ],
    "F1": [f"{meanTableF1[i]:.3f} ± {stdTableF1[i]:.3f}" for i in range(3)],
    "ROC-AUC": [f"{meanTableRocAuc[i]:.3f} ± {stdTableRocAuc[i]:.3f}" for i in range(3)]
})

print(finalComparisonTable)

                      Model       Accuracy      Precision         Recall  \
0           DummyClassifier  0.902 ± 0.005  0.000 ± 0.000  0.000 ± 0.000   
1       Logistic Regression  0.897 ± 0.012  0.313 ± 0.270  0.129 ± 0.112   
2  Random Forest Classifier  0.882 ± 0.009  0.000 ± 0.000  0.000 ± 0.000   

              F1        ROC-AUC  
0  0.000 ± 0.000  0.500 ± 0.000  
1  0.179 ± 0.153  0.803 ± 0.052  
2  0.000 ± 0.000  0.790 ± 0.099  


## Day 5 Findings

The DummyClassifier achieved about 90.2% accuracy, but it had zero precision, recall, and F1 with a ROC-AUC of 0.5. This shows that the high accuracy comes from predicting the majority class while failing to detect defective modules. Logistic Regression performed better for defect detection, with a recall of 0.129, F1-score of 0.179, and ROC-AUC of 0.803.
Random Forest achieved lower accuracy than the DummyClassifier and Logistic Regression. Random Forest produced zero recall and F1 at the current classification threshold, meaning it did not correctly classify any defective modules during the five validation folds. However, its ROC-AUC of about 0.79 suggests that its probability scores still contained discriminatory information. Therefore, Logistic Regression was stronger overall for defect detection.
